### Importing & installing neccessary packages

In [1510]:
import pandas as pd
# pip install openpyxl

## Loading and exploring the raw Excel files

Before selecting the correct data to clean and harmonize, we first explore the Excel files to understand their structure.  
Each file contains several sheets, so this initial step helps us verify which one holds the program data.  

(This is an exploratory step - we won't use these variables later, since we will reload only the relevant sheets).

In [1511]:
# Commenting out these lines of code after deciding which sheet to work with.
# df_20 = pd.read_excel("resultat-ansokningsomgang-2020.xlsx")
# df_21 = pd.read_excel("resultat-ansokningsomgang-2021.xlsx")
# df_22 = pd.read_excel("resultat-ansokningsomgang-2022.xlsx")

print("Exploration step only — main datasets will be loaded later as df_2020, df_2021, df_2022 (Tabell 1).")

Exploration step only — main datasets will be loaded later as df_2020, df_2021, df_2022 (Tabell 1).


In [1512]:
# To have a broader understanding of the dataset, we set "None" to the max rows, columns, width and max colwith. Otherwise the data gets truncated per default.
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', None)

### Before cleaning the data, we need take a look on the file structure. Each Excel file contains several sheets (or tabs), and we need to identify which one holds the relevant data that we can work with.


### We start analyzing the 2020 file and contiune with the 2021 & 2022 file:

In [1513]:
# First we check how many tabs the file contains.
# Create a variable to easier read information
path_2020 = "resultat-ansokningsomgang-2020.xlsx"

xls_20 = pd.ExcelFile(path_2020)
xls_20.sheet_names

['Innehållsförteckning',
 'Definitioner och förklaringar',
 'Tabell 1',
 'Tabell 2',
 'Tabell 3',
 'Tabell 4']

In [1514]:
# We create a variable for each tab and use len to check the length of each tab.
# Conclusion: Tab 1 & 2 contains the same ammount of rows (484), while tab 3 & 4 contains 1482 and 1903.
t20_schema  = pd.read_excel(path_2020, sheet_name="Innehållsförteckning")
t20_1 = pd.read_excel(path_2020, sheet_name="Tabell 1")
t20_2 = pd.read_excel(path_2020, sheet_name="Tabell 2")
t20_3 = pd.read_excel(path_2020, sheet_name="Tabell 3")
t20_4 = pd.read_excel(path_2020, sheet_name="Tabell 4")

print("Tab 1:", len(t20_1))
print("Tab 2:", len(t20_2))
print("Tab 3:", len(t20_3))
print("Tab 4:", len(t20_4))

Tab 1: 484
Tab 2: 484
Tab 3: 1482
Tab 4: 1903


In [1515]:
# We use .head() and .columns to check why there's a bigger set of rows in the last tab.
# We also check the other tabs (2, 3 and 4).
t20_1.head()
t20_1.columns

Index(['Utbildningsområde', 'Utbildningsnamn', 'Län', 'Kommun',
       'Flera studiekommuner', 'Antal kommuner', 'Antal län', 'YH-poäng',
       'Utbildningsanordnare administrativ enhet', 'Diarienummer'],
      dtype='object')

### Analyzing 2021 file:

In [1516]:
path_2021 = "resultat-ansokningsomgang-2021.xlsx"

xls_21 = pd.ExcelFile(path_2021)
xls_21.sheet_names

['Innehållsförteckning',
 'Definitioner och förklaringar',
 'Tabell 1',
 'Tabell 2',
 'Tabell 3',
 'Tabell 4']

In [1517]:
# Conclusion: Tab 1 & 2 contains the same ammount of rows (426), while tab 3 & 4 contains 1238 and 1586.
t21_schema  = pd.read_excel(path_2021, sheet_name="Innehållsförteckning")
t21_1 = pd.read_excel(path_2021, sheet_name="Tabell 1")
t21_2 = pd.read_excel(path_2021, sheet_name="Tabell 2")
t21_3 = pd.read_excel(path_2021, sheet_name="Tabell 3")
t21_4 = pd.read_excel(path_2021, sheet_name="Tabell 4")

print("Tab 1:", len(t21_1))
print("Tab 2:", len(t21_2))
print("Tab 3:", len(t21_3))
print("Tab 4:", len(t21_4))

Tab 1: 426
Tab 2: 426
Tab 3: 1238
Tab 4: 1586


In [1518]:
# Again - we check all the tabs (2, 3 & 4 aswell).
t21_1.head()
t21_1.columns

Index(['Utbildningsområde', 'Utbildningsnamn', 'Län', 'Kommun',
       'Flera studiekommuner', 'Antal kommuner', 'Antal län', 'YH-poäng',
       'Utbildningsanordnare administrativ enhet', 'Diarienummer'],
      dtype='object')

### Analyzing the 2022 file:

In [1519]:
path_2022 = "resultat-ansokningsomgang-2022.xlsx"

xls_22 = pd.ExcelFile(path_2022)
xls_22.sheet_names

['Innehållsförteckning',
 'Definitioner och förklaringar',
 'Tabell 1',
 'Tabell 2',
 'Tabell 3',
 'Tabell 4']

In [1520]:
# Conclusion: Tab 1 & 2 contains the same ammount of rows (420), while tab 3 & 4 contains 1207 and 1621.
t22_schema  = pd.read_excel(path_2022, sheet_name="Innehållsförteckning")
t22_1 = pd.read_excel(path_2022, sheet_name="Tabell 1")
t22_2 = pd.read_excel(path_2022, sheet_name="Tabell 2")
t22_3 = pd.read_excel(path_2022, sheet_name="Tabell 3")
t22_4 = pd.read_excel(path_2022, sheet_name="Tabell 4")

print("Tab 1:", len(t22_1))
print("Tab 2:", len(t22_2))
print("Tab 3:", len(t22_3))
print("Tab 4:", len(t22_4))

Tab 1: 420
Tab 2: 420
Tab 3: 1207
Tab 4: 1621


In [1521]:
# Again - we check all the tabs (2, 3 & 4 aswell).
t22_1.head()
t22_1.columns

Index(['Utbildningsområde', 'Utbildningsnamn', 'Län', 'Kommun', 'Diarienummer',
       'Flera kommuner', 'Antal kommuner', 'YH-poäng',
       'Utbildningsanordnare administrativ enhet'],
      dtype='object')

# Conclusion:

### Table 1 & 2 have the same number of rows in each file, (the same approved programs, just in a different order).
### Table 3 & 4 have more rows, this is because they contain details per round/location, which in turn created more sets of rows.

### Therefore, I will use Table 1 as the master for harmonization for each file.

Even though the Excel files contain several tables, I decided to use Tabell 1 as the main dataset.
Tables 2–4 contain either duplicates or more detailed breakdowns (per place or decision round).
Merging them would create duplicate program entries and make the harmonization messy.

Instead, I will use Tabell 1 for all years and optionally merge specific details from other tables later if needed.




### Loading "Tabell 1" from each file:

In [1522]:
df_2020 = pd.read_excel("resultat-ansokningsomgang-2020.xlsx", sheet_name="Tabell 1")
df_2021 = pd.read_excel("resultat-ansokningsomgang-2021.xlsx", sheet_name="Tabell 1")
df_2022 = pd.read_excel("resultat-ansokningsomgang-2022.xlsx", sheet_name="Tabell 1")

In [1523]:
# Using head to see a simple overview of the columns.
# Our 2020 & 2021 datasets both contains equal number of columns (10), and the column names are identical.
# However the 2022 dataset has 9 columns and their names are slightly different.

#df_2020.head()
#df_2021.head()
df_2022.head()

,Utbildningsområde,Utbildningsnamn,Län,Kommun,Diarienummer,Flera kommuner,Antal kommuner,YH-poäng,Utbildningsanordnare administrativ enhet
0,Data/IT,.NET Utvecklare,Gävleborg,Gävle,MYH 2022/4476,Nej,1,400,Plushögskolan AB - Teknikhögskolan
1,Data/IT,Cloud & ICT Engineer,Stockholm,Stockholm,MYH 2022/5253,Nej,1,400,Lernia Utbildning AB
2,Data/IT,Data Engineer,Stockholm,Stockholm,MYH 2022/4977,Nej,1,400,Stiftelsen Stockholms Tekniska Institut
3,Data/IT,Data Manager,Stockholm,Stockholm,MYH 2022/5462,Nej,1,400,TUC Sweden AB - Yrkeshögskola
4,Data/IT,DevOps Engineer,Stockholm,Stockholm,MYH 2022/5257,Nej,1,400,Lernia Utbildning AB


### Using `.info()` method to inspect structure of the datasets for each year.

This helps us confirm the number of rows, columns, and data types once again - better to be safe than sorry!

After inspecting the datasets, we can verify that all three years contain similar information (with some exceptions (these will be mentioned in the comments below)).

In [1524]:
# To get a more detailed overview, we use .info() method on each dataset.
# This shows us how many rows and columns there is, the column datatypes and if there's any missing values in any columns.
print("2020")
df_2020.info()

2020
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 484 entries, 0 to 483
Data columns (total 10 columns):
 #   Column                                    Non-Null Count  Dtype 
---  ------                                    --------------  ----- 
 0   Utbildningsområde                         484 non-null    object
 1   Utbildningsnamn                           484 non-null    object
 2   Län                                       484 non-null    object
 3   Kommun                                    484 non-null    object
 4   Flera studiekommuner                      484 non-null    object
 5   Antal kommuner                            484 non-null    int64 
 6   Antal län                                 484 non-null    int64 
 7   YH-poäng                                  484 non-null    int64 
 8   Utbildningsanordnare administrativ enhet  484 non-null    object
 9   Diarienummer                              484 non-null    object
dtypes: int64(3), object(7)
memory usage: 37.9+ KB

In [1525]:
print("2021")
df_2021.info()

2021
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426 entries, 0 to 425
Data columns (total 10 columns):
 #   Column                                    Non-Null Count  Dtype 
---  ------                                    --------------  ----- 
 0   Utbildningsområde                         426 non-null    object
 1   Utbildningsnamn                           426 non-null    object
 2   Län                                       426 non-null    object
 3   Kommun                                    426 non-null    object
 4   Flera studiekommuner                      426 non-null    object
 5   Antal kommuner                            426 non-null    int64 
 6   Antal län                                 426 non-null    int64 
 7   YH-poäng                                  426 non-null    int64 
 8   Utbildningsanordnare administrativ enhet  426 non-null    object
 9   Diarienummer                              426 non-null    object
dtypes: int64(3), object(7)
memory usage: 33.4+ KB

In [1526]:
print("2022")
df_2022.info()

2022
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 420 entries, 0 to 419
Data columns (total 9 columns):
 #   Column                                    Non-Null Count  Dtype 
---  ------                                    --------------  ----- 
 0   Utbildningsområde                         420 non-null    object
 1   Utbildningsnamn                           420 non-null    object
 2   Län                                       420 non-null    object
 3   Kommun                                    420 non-null    object
 4   Diarienummer                              420 non-null    object
 5   Flera kommuner                            420 non-null    object
 6   Antal kommuner                            420 non-null    int64 
 7   YH-poäng                                  420 non-null    int64 
 8   Utbildningsanordnare administrativ enhet  420 non-null    object
dtypes: int64(2), object(7)
memory usage: 29.7+ KB


### Using the `tolist()` method to convert the Pandas Series into a plain Python list, so it can help us quickly view the column names from each dataset and spot small differances.

In [1527]:
print("2020:", df_2020.columns.tolist())
print("2021:", df_2021.columns.tolist())
print("2022:", df_2022.columns.tolist())

2020: ['Utbildningsområde', 'Utbildningsnamn', 'Län', 'Kommun', 'Flera studiekommuner', 'Antal kommuner', 'Antal län', 'YH-poäng', 'Utbildningsanordnare administrativ enhet', 'Diarienummer']
2021: ['Utbildningsområde', 'Utbildningsnamn', 'Län', 'Kommun', 'Flera studiekommuner', 'Antal kommuner', 'Antal län', 'YH-poäng', 'Utbildningsanordnare administrativ enhet', 'Diarienummer']
2022: ['Utbildningsområde', 'Utbildningsnamn', 'Län', 'Kommun', 'Diarienummer', 'Flera kommuner', 'Antal kommuner', 'YH-poäng', 'Utbildningsanordnare administrativ enhet']


### After analyzing the datasets with `.info()` and `.tolist()` methods:

We can observe that we're missing the column "Antal län", which both the dataset 2020 and 2021 have.
The column name "Flera studiekommuner" (which both 2020 & 2021 has), is named "Flera kommuner" in this dataset, but does overall give the same information (it's just a different column name).

Conclusion: One missing column (Antal län) and one column named different (Flera kommuner).
Besides the above stated, all other information looks good (there's no missing values and all the datatypes are either objects or int64).

### It makes sense to rename the columns so they all have the same structure.
- Renaming: "Flera kommuner" to "Flera studiekommuner".

### Furthermore we need to change one datatype:
- In dataset 2022 the "Flera kommuner" column the datatype is set to object, meanwhile the other two datasets the datatype is int64, so later on when cleaning the data we'll rewrite it into an int64 aswell.

### Before we contiune making changes to the datasets, we want to check if there's anything that sticks out using the `.nunique()` method.

This returns the number of unique values in each column, which we can analyze.

In [1528]:
# Using a for-loop to print out the number of values in each column.
for col in df_2020.columns:
    print(f"{col}: {df_2020[col].nunique()} unique values")

Utbildningsområde: 15 unique values
Utbildningsnamn: 380 unique values
Län: 21 unique values
Kommun: 89 unique values
Flera studiekommuner: 2 unique values
Antal kommuner: 7 unique values
Antal län: 5 unique values
YH-poäng: 39 unique values
Utbildningsanordnare administrativ enhet: 156 unique values
Diarienummer: 484 unique values


In [1529]:
for col in df_2021.columns:
    print(f"{col}: {df_2021[col].nunique()} unique values")

Utbildningsområde: 15 unique values
Utbildningsnamn: 352 unique values
Län: 21 unique values
Kommun: 89 unique values
Flera studiekommuner: 2 unique values
Antal kommuner: 7 unique values
Antal län: 5 unique values
YH-poäng: 38 unique values
Utbildningsanordnare administrativ enhet: 167 unique values
Diarienummer: 426 unique values


In [1530]:
for col in df_2022.columns:
    print(f"{col}: {df_2022[col].nunique()} unique values")

Utbildningsområde: 15 unique values
Utbildningsnamn: 348 unique values
Län: 21 unique values
Kommun: 83 unique values
Diarienummer: 420 unique values
Flera kommuner: 2 unique values
Antal kommuner: 6 unique values
YH-poäng: 45 unique values
Utbildningsanordnare administrativ enhet: 162 unique values


### Analyzing the results we got from the for-loop with `.nunique()` method:

- Utbildningsområde: 15 unique values

- Utbildningsnamn: ~348–352 unique values

- Län and Kommun: Län: 21 unique values every year, which matches Sweden’s 21 counties.

- Flera studiekommuner: Flera studiekommuner only has 2 value.

- Antal kommuner/län: Antal kommuner and Antal län have very few unique values (5–7).

- YH-poäng: Shows program lengths, e.g., 225, 400, 430, 500, etc.

- Utbildningsanordnare administrativ enhet - About 160–167 unique organizers.

- Diarienummer - Each program has a unique ID.

**Summary**: The only thing that should catch our attention is the "Diarienummer".  
The datasets of 2021 (rows: 426, unique values: 426) and 2022 (rows: 420, unique values: 420) match.  
However, year 2020 (rows: 420, unique values: 484) means that some diarienummer (which should be unique) appear more than once.

### Analyzing and cleaning the column "Diarienummer" in 2020 dataset with the `.duplicated()` and `.sum()` method.

The .duplicated() checks if any "Diarienummer" values appear more than once.
The .sum() method then counts how many duplicated entries exist in total.

The result "np.int64(0)" means there are no duplicated "Diarienummer" values in the 2020 dataset.
Therefore, all program IDs are unique. The small difference between total rows (484)
and unique values (420) might be caused by missing IDs or formatting errors (whitespace or such).

Conclusion: We dont have to manipulate the data.

In [1531]:
df_2020["Diarienummer"].duplicated().sum()

np.int64(0)

### Analyzing missing values
We'll use the `.isna().sum()` method to count any missing data in each dataset.  
All three dataframes (2020, 2021, 2022) shows **0 missing values** in every column,  
which means the datasets are complete and no further cleaning is required for missing data.

In [1532]:
df_2020.isna().sum()

Utbildningsområde                           0
Utbildningsnamn                             0
Län                                         0
Kommun                                      0
Flera studiekommuner                        0
Antal kommuner                              0
Antal län                                   0
YH-poäng                                    0
Utbildningsanordnare administrativ enhet    0
Diarienummer                                0
dtype: int64

In [1533]:
df_2021.isna().sum()

Utbildningsområde                           0
Utbildningsnamn                             0
Län                                         0
Kommun                                      0
Flera studiekommuner                        0
Antal kommuner                              0
Antal län                                   0
YH-poäng                                    0
Utbildningsanordnare administrativ enhet    0
Diarienummer                                0
dtype: int64

In [1534]:
df_2022.isna().sum()

Utbildningsområde                           0
Utbildningsnamn                             0
Län                                         0
Kommun                                      0
Diarienummer                                0
Flera kommuner                              0
Antal kommuner                              0
YH-poäng                                    0
Utbildningsanordnare administrativ enhet    0
dtype: int64

No missing values found.

# Cleaning data - Now that we've inspected and analyzed the datasets, it's time to clean the data.

Rename the columns so they all have the same structure.
- Renaming the 2022 column: "Flera kommuner" to "Flera studiekommuner".

Change one datatype:
- In dataset 2022 the "Flera kommuner" column the datatype is set to object, meanwhile the other two datasets the datatype is int64, so later on when cleaning the data we'll rewrite it into an int64 aswell.

Add one more column:
- To make the datasets as identical as possible, we'll add "Antal län" to the 2022 dataset (which was missing as we saw in the analyze section).

In [1535]:
# To be on the safe side, we use the .tolist() method again before cleaning the relevant data.
print("2020:", df_2020.columns.tolist())
print("2021:", df_2021.columns.tolist())
print("2022:", df_2022.columns.tolist())

2020: ['Utbildningsområde', 'Utbildningsnamn', 'Län', 'Kommun', 'Flera studiekommuner', 'Antal kommuner', 'Antal län', 'YH-poäng', 'Utbildningsanordnare administrativ enhet', 'Diarienummer']
2021: ['Utbildningsområde', 'Utbildningsnamn', 'Län', 'Kommun', 'Flera studiekommuner', 'Antal kommuner', 'Antal län', 'YH-poäng', 'Utbildningsanordnare administrativ enhet', 'Diarienummer']
2022: ['Utbildningsområde', 'Utbildningsnamn', 'Län', 'Kommun', 'Diarienummer', 'Flera kommuner', 'Antal kommuner', 'YH-poäng', 'Utbildningsanordnare administrativ enhet']


### Renaming 2022 column name with `.rename()` method. We'll also save the changes directly to the DataFrame with the `inplace=True` parameter.
### After renaming it, we convert it from object to int64 with the `astype("int64")` method and save it.
Saving it with `.astype()` as it creates and returns a new copy of the DataFrame with the new datatype. Earlier we used `inplace=True`, but it doesn't work here since `.astype()` doesn't have an `inplace=True` parameter.
Therefore, we'll create a new assignment to store the updated column back into the DataFrame.

But before we change it to an int64, we need to know why it's an object, that's why we check it with the `.unique()` method.

In [1536]:
df_2022.rename(columns={"Flera kommuner": "Flera studiekommuner"}, inplace=True)

In [1537]:
# Here we see that the question was put as a "Yes" or "No" question. This means that we can convert the "Yes" answers to "1" and no to "0" to safely convert it to an int64 type.
df_2022["Flera studiekommuner"].unique()

array(['Nej', 'Ja'], dtype=object)

In [1538]:
# Here we use the .replace to alter the "Ja" to "1" and the "Nej" to "0", then save it with the astype method.
# We'll also use the infer_object() to future proof the code, otherwise Pandas future version might not convert the comumn type automatically from object to int64, like it does now.
df_2022["Flera studiekommuner"] = (df_2022["Flera studiekommuner"].replace({"Ja": 1, "Nej": 0}).infer_objects(copy=False).astype("int64"))

# About the FutureWarning: It seems that Pandas will eventually remove the .replace() method - which means that it won’t automatically convert the column type from object → int64 like it does now.
# I've played around with the code but cant seem to be able to navigate around the futurewarning error.
# For now - it works, I would really want to try to fix this but im having difficulty doing so at the moment. Hopefully, with more practice I'll get better at working with Pandas.

C:\Users\Elias\AppData\Local\Temp\ipykernel_5400\2053902474.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_2022["Flera studiekommuner"] = (df_2022["Flera studiekommuner"].replace({"Ja": 1, "Nej": 0}).infer_objects(copy=False).astype("int64"))


In [1539]:
# Sanity check - was the conversion successfull? - Yes.
df_2022["Flera studiekommuner"].unique()

array([0, 1])

### Checking the "Antal län" in the datasets:
- Using the `.unique()` method to verify where it's missing (we can see in the code below that it's the 2022 dataset that's missing the column)

- Adding the missing column in dataset 2022 by giving it a default value, giving the datasets a structure consistency. If we didn't add this column, the 2022 file would have a missing column, which means that Pandas would fill it with NaN (missing values).

- To know what value we need to enter into the new column, we use the `.value_counts()` method in the 2020 and 2021 datasets to check what the most common input is.

In [1540]:
df_2020["Antal län"].unique()

array([1, 2, 3, 5, 4])

In [1541]:
df_2021["Antal län"].unique()

array([1, 2, 3, 4, 5])

In [1542]:
# In this dataset we can observe that the column is missing. (We're commenting it out so we can toggle the "Run All" button without getting any errors).
#df_2022["Antal län"].unique()

In [1543]:
df_2020["Antal län"].value_counts()

Antal län
1    445
2     25
3     10
5      2
4      2
Name: count, dtype: int64

In [1544]:
df_2021["Antal län"].value_counts()

Antal län
1    390
2     26
3      8
4      1
5      1
Name: count, dtype: int64

Considering that both datasets above has a huge majority in their array value 1, we'll set the 2022 column to equal 1.

Note: We could set the column to show missing data (with = np.nan), but after giving it some thought, it seems more appropriate to set it to 1 due to the information stated above.

In [1545]:
df_2022["Antal län"] = 1
df_2022["Antal län"].unique() # Sanity check - See if the column has been added successfully - Yes.

array([1])

### Keeping the columns that describe the education programs: 
- Utbildningsområde
- Utbildningsnamn
- Län
- Kommun
- Flera studiekommuner
- Antal kommuner
- Diarienummer
- YH-poäng
- Utbildningsanordnare administrativ enhet

### Note: "Antal län" doesn't give us much information in afterthought, so we'll remove it. 
We'll keep the previously steps we've made as a reminder that sometimes changes must be made, even after carefully going over data.


We'll be using a list variable to store all the relevant columns, then we'll apply the changes with the same variable.

In [1546]:
# Save the relevant columns to variable "list". Keep in mind that we're not adding the "Antal län" in this list - which removes it.
columns_to_keep = [
    "Utbildningsområde",
    "Utbildningsnamn",
    "Län",
    "Kommun",
    "Flera studiekommuner",
    "YH-poäng",
    "Utbildningsanordnare administrativ enhet",
    "Diarienummer"
]

# Sanity check
print(columns_to_keep)

['Utbildningsområde', 'Utbildningsnamn', 'Län', 'Kommun', 'Flera studiekommuner', 'YH-poäng', 'Utbildningsanordnare administrativ enhet', 'Diarienummer']


In [1547]:
# Save the changes.
df_2020 = df_2020[columns_to_keep]
df_2021 = df_2021[columns_to_keep]
df_2022 = df_2022[columns_to_keep]

In [1548]:
# Sanity check: .equal() method gives us a boolean expression. If the columns are identical, it returns True, otherwise it returns False. 
# It returns True, so we've successfully converted the columns.
print(df_2020.columns.equals(df_2021.columns))
print(df_2020.columns.equals(df_2022.columns))

True
True


### Enrichment: It also makes sense to add a "Year" column and change the language from swedish to english.

We'll do that using the same methods as previously in the notebook.
Keeping the changes with the `inplace=True` parameter.
And then do a check if the changes was successful.

In [1549]:
# Adding year column to datasets.
df_2020["Year"] = 2020
df_2021["Year"] = 2021
df_2022["Year"] = 2022

In [1550]:
# Save the relevant columns to variable "list".
column_translation = {
    "Utbildningsområde": "Education Field",
    "Utbildningsnamn": "Program Name",
    "Län": "County",
    "Kommun": "Municipality",
    "Flera studiekommuner": "Multiple Municipalities",
    "Antal kommuner": "Number of Municipalities",
    "Antal län": "Number of Counties",
    "YH-poäng": "YH Points",
    "Utbildningsanordnare administrativ enhet": "Provider Admin Unit",
    "Diarienummer": "Case Number"
}

In [1551]:
# Since all the datasets now have identical columns, we can apply it easily to all three using the .rename() method.
df_2020.rename(columns=column_translation, inplace=True)
df_2021.rename(columns=column_translation, inplace=True)
df_2022.rename(columns=column_translation, inplace=True)

In [1552]:
# Sanity check.
print("2020:", df_2020.columns.tolist())
print("2021:", df_2021.columns.tolist())
print("2022:", df_2022.columns.tolist()),

# Another sanity check.
print(df_2020.columns.equals(df_2021.columns))
print(df_2020.columns.equals(df_2022.columns))

2020: ['Education Field', 'Program Name', 'County', 'Municipality', 'Multiple Municipalities', 'YH Points', 'Provider Admin Unit', 'Case Number', 'Year']
2021: ['Education Field', 'Program Name', 'County', 'Municipality', 'Multiple Municipalities', 'YH Points', 'Provider Admin Unit', 'Case Number', 'Year']
2022: ['Education Field', 'Program Name', 'County', 'Municipality', 'Multiple Municipalities', 'YH Points', 'Provider Admin Unit', 'Case Number', 'Year']
True
True


### Making sure that that there's no inconsictencies in the text by trimming and capetalizing the column names, we'll also convert them to string types:
- Using the `.astype(str)` to convert the column to string type, even if it previously contained mixed types (like text + NaN), which means that the cells that are `None` or `integers` becomes "None" and "numbers" (stringtype).
- Using the `.str.strip()` to remove whitespaces, tabs or newlines.
- Using the `.str.title()` to capatalize column names.

This is usefull when working with Streamlit, otherwise the filters might split e.g. "Stockholm" and "Stockholm " to two different sections.

Note: We're only passing these changes to the relevant columns - "Multiple Municipalities", "Number of Municipalities", "Number of Counties" and "YH Points" are all numerical types, so they need no change.
It's also worth mentioning that the "Case Number" is sort of a identifier, and should by default be "right" from the start - So we should not have to make any changes to it.

In [1553]:
# Using a for-loop for relevant columns:
for df in [df_2020, df_2021, df_2022]:
    for col in ["Program Name", "Education Field", "County", "Municipality", "Provider Admin Unit"]:
        df[col] = df[col].astype(str).str.strip().str.title()

In [1554]:
# Sanity check for datatypes.
df_2020.dtypes

Education Field            object
Program Name               object
County                     object
Municipality               object
Multiple Municipalities    object
YH Points                   int64
Provider Admin Unit        object
Case Number                object
Year                        int64
dtype: object

In [1555]:
# Sanity check for datatypes.
df_2021.dtypes

Education Field            object
Program Name               object
County                     object
Municipality               object
Multiple Municipalities    object
YH Points                   int64
Provider Admin Unit        object
Case Number                object
Year                        int64
dtype: object

In [1556]:
# Sanity check for datatypes.
df_2022.dtypes

Education Field            object
Program Name               object
County                     object
Municipality               object
Multiple Municipalities     int64
YH Points                   int64
Provider Admin Unit        object
Case Number                object
Year                        int64
dtype: object

### While doing a sanity check in the datatypes we've encountered an anomaly in the "Multiple Municipalities" in the datasets from year 2020 and 2021.

They are both currently the datatype object. Since it's a Yes/No question, it makes sence to convert them to a numeric datatype - which we'll do in the following cells (as we did with the 2022 dataset).

In [1557]:
# Convert from 'Ja'/'Nej' to 1/0 for df_2020
df_2020["Multiple Municipalities"] = (df_2020["Multiple Municipalities"].replace({"Ja": 1, "Nej": 0}).astype("int64"))

C:\Users\Elias\AppData\Local\Temp\ipykernel_5400\3668130375.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_2020["Multiple Municipalities"] = (df_2020["Multiple Municipalities"].replace({"Ja": 1, "Nej": 0}).astype("int64"))


In [1558]:
# Convert from 'Ja'/'Nej' to 1/0 for df_2021
df_2021["Multiple Municipalities"] = (df_2021["Multiple Municipalities"].replace({"Ja": 1, "Nej": 0}).astype("int64"))

C:\Users\Elias\AppData\Local\Temp\ipykernel_5400\1052299251.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_2021["Multiple Municipalities"] = (df_2021["Multiple Municipalities"].replace({"Ja": 1, "Nej": 0}).astype("int64"))


In [1559]:
# Sanity check - conversion successful.
df_2020.dtypes

Education Field            object
Program Name               object
County                     object
Municipality               object
Multiple Municipalities     int64
YH Points                   int64
Provider Admin Unit        object
Case Number                object
Year                        int64
dtype: object

In [1560]:
# Sanity check - conversion successful.
df_2021.dtypes

Education Field            object
Program Name               object
County                     object
Municipality               object
Multiple Municipalities     int64
YH Points                   int64
Provider Admin Unit        object
Case Number                object
Year                        int64
dtype: object

In [1561]:
# Sanity check - Everything looks fine now.
df_2022.dtypes

Education Field            object
Program Name               object
County                     object
Municipality               object
Multiple Municipalities     int64
YH Points                   int64
Provider Admin Unit        object
Case Number                object
Year                        int64
dtype: object

### Enrichment: Taking the YH Points column in considuration, it would be be able to see how long it would take to achive the YH points in the cells.

Each program’s duration can be estimated using a common rule from MYH:

5 YH-points ≈ 1 study week.
So, for example, a 400 YH-point program lasts about 80 weeks (400 ÷ 5). Using this rule we can calculate the duration in weeks and years.

In [1562]:
# Using a for-loop to go through the datasets.
# Calculate study duration in weeks (5 YH points = 1 study week), and save it to the new made column. Also rounding weeks to whole numbers (int64).
# Convert weeks to years (about 52 weeks per year), and save it to the new made column. Also rounding years to .2 decimals (float).
for df in [df_2020, df_2021, df_2022]:
    df["Study Weeks"] = (df["YH Points"] / 5).round().astype("Int64")
    df["Study Years"] = (df["Study Weeks"] / 52).round(2)

In [1563]:
# Sanity check: Does the columns exist? Does the Value make sense? - Yes.
# To not clutter the screen, we've used the three datasets in same cell, here we have to comment them in and out to check them manually.

df_2020[["YH Points", "Study Weeks", "Study Years"]].head()
df_2021[["YH Points", "Study Weeks", "Study Years"]].head()
df_2022[["YH Points", "Study Weeks", "Study Years"]].head()


,YH Points,Study Weeks,Study Years
0,400,80,1.54
1,400,80,1.54
2,400,80,1.54
3,400,80,1.54
4,400,80,1.54


In [1564]:
# Sanity check: Does the columns have right datatypes? - Yes.
# To not clutter the screen, we've used the three datasets in same cell, here we have to comment them in and out to check them manually.

df_2020[["Study Weeks", "Study Years"]].dtypes
df_2021[["Study Weeks", "Study Years"]].dtypes
df_2022[["Study Weeks", "Study Years"]].dtypes


Study Weeks      Int64
Study Years    Float64
dtype: object

In [1565]:
# Sanity check: A quick statistical summary, does it make sense? - Yes.
# To not clutter the screen, we've used the three datasets in same cell, here we have to comment them in and out to check them manually.

df_2020[["Study Weeks", "Study Years"]].describe()
#df_2021[["Study Weeks", "Study Years"]].describe()
#df_2022[["Study Weeks", "Study Years"]].describe()


,Study Weeks,Study Years
count,484.0,484.0
mean,70.134298,1.349236
std,18.047572,0.347331
min,20.0,0.38
25%,60.0,1.15
50%,80.0,1.54
75%,80.0,1.54
max,120.0,2.31


# We've analyzed, cleaned and made some enrichment to our data, we'll move on to the finall step in this notebook - Harmonization.

# Harmonization:

We're gonna harmonize column order using the 2020 as the template and then concatinate them.

In [ ]:
# We create variable that holds the 2020 columns, add those to the 2021 datasets columns, and add those columns to the 2022 dataset columns.
cols = df_2020.columns
df_2021 = df_2021[cols]
df_2022 = df_2022[cols]

# We create a variable that stores the result of the concatination.
# We use the pd.concat() method that combines our DataFrames.
# We use the ignore_index=True to ignore the original index vaules from each DataFrame and create new indexes for the new combined DataFrame - else we'll get duplicates or conflicting rows indices.
combined = pd.concat([df_2020, df_2021, df_2022], ignore_index=True)

# Check the .shape - which returns the number of rows and number of columns as a tuple
print(combined.shape) # output: (1330, 11)
combined.head()


# Check if there's any duplicates and sum them up
print("Duplicate Case Numbers:", combined["Case Number"].duplicated().sum()) # output: Duplicate Case Numbers: 0

(1330, 11)
Duplicate Case Numbers: 0


In [ ]:
# Check .sample(50) method to see 50 random rows of the combined dataset looks like (we can run the cell serveral times to check if there's any anomalys).
combined.sample(50)

In [ ]:
# Quick sanity check: Is there any Nan-values? - No.
print(combined.isna().sum())

Education Field            0
Program Name               0
County                     0
Municipality               0
Multiple Municipalities    0
YH Points                  0
Provider Admin Unit        0
Case Number                0
Year                       0
Study Weeks                0
Study Years                0
dtype: int64


### We got our 11 columns, all the cells in the columns looks fine - no obvious anomalys, index labels also looks fine and there's no duplicates or Nan-values.
### Saving the combined DataFrames.

In [1568]:
# Save the combined DataFrames to a .csv-file.
# Using the index=False here aswell, else we're gonna end up with a default indexs in the first column (.csv files creates them, index=False removes them).
combined.to_csv("harmonized_yh_2020_2022.csv", index=False)